src/compiler/checker.ts

```diff
@@ -4420,12 +4420,26 @@ namespace ts {
             let container = getThisContainer(node, /*includeArrowFunctions*/ false);
             let parent = container && container.parent;
             if (parent && (isClassLike(parent) || parent.kind === SyntaxKind.InterfaceDeclaration)) {
-                if (!(container.flags & NodeFlags.Static)) {
+                if (!(container.flags & NodeFlags.Static) && !isConstructorParameter(node, container)) {
                     return getDeclaredTypeOfClassOrInterface(getSymbolOfNode(parent)).thisType;
                 }
             }
             error(node, Diagnostics.A_this_type_is_available_only_in_a_non_static_member_of_a_class_or_interface);
             return unknownType;
+
+            function isConstructorParameter(node: Node, container: Node) {

```

Comment:
That kind of optimization isn't technically part of any JS spec, but that kind of inlining is what you'd expect v8's [crankshaft](http://blog.chromium.org/2010/12/new-crankshaft-for-v8.html) optimizing compiler ([and now the TurboFan JIT](http://v8project.blogspot.com/2015/07/digging-into-turbofan-jit.html)) to do. You can't really know the performance profile against a given version of v8 without testing though, so it can be better to write slightly more optimized code to begin with, and avoid all the testing, knowing you've already done your best, rather than leaning on the compiler implementation.

[For example](https://code.google.com/p/v8/issues/detail?id=3842), the reason nobody likes to use node 0.12 is because the v8 version within it had a perf regression whereby the opimizing compiler failed to optimize visitor-dispatch type patterns (`obj[action]();`), lowing their performance significantly compared to, say, a switch statement and non-variable lookups. This has since been fixed and the perf of the two is roughly equal again, but it just goes to show how little you can take for granted, and how much we rely on the compiler for speed.